source/Octopus.Server.Client/Operations/RegisterMachineOperation.cs

```diff
@@ -119,22 +125,62 @@ void ValidateTenantTags(IOctopusSpaceRepository repository)
                 return;
 
             var tagSets = repository.TagSets.FindAll();
-            var missingTags = TenantTags.Where(tt => !tagSets.Any(ts => ts.Tags.Any(t => t.CanonicalTagName.Equals(tt, StringComparison.OrdinalIgnoreCase)))).ToList();
+            var missingTags = TenantTags.Where(tt =>
+                    !tagSets.Any(ts =>
+                        ts.Tags.Any(t => t.CanonicalTagName.Equals(tt, StringComparison.OrdinalIgnoreCase))))
+                .ToList();
 
             if (missingTags.Any())
-                throw new InvalidRegistrationArgumentsException(CouldNotFindMessage("tag", missingTags.ToArray()));
+                throw new InvalidRegistrationArgumentsException(CouldNotFindByNameMessage("tag",
+                    missingTags.ToArray()));
         }
 
         List<EnvironmentResource> GetEnvironments(IOctopusSpaceRepository repository)
         {
-            var selectedEnvironments = repository.Environments.FindByNames(EnvironmentNames);
+            List<EnvironmentResource> environments = new();
+            if (EnvironmentNames is not null && EnvironmentNames.Any())
+            {
+                var envsByName = repository.Environments.FindByNames(EnvironmentNames);
+                environments.AddRange(envsByName);
 
-            var missing = EnvironmentNames.Except(selectedEnvironments.Select(e => e.Name), StringComparer.OrdinalIgnoreCase).ToList();
+                //if there are any missing environment names only, then we want to throw an exception and not check the missing names against slugs or ids
+                var missingByNameOnly = EnvironmentNames
+                    .Except(envsByName.Select(e => e.Name), StringComparer.OrdinalIgnoreCase).ToArray();
 
-            if (missing.Any())
-                throw new InvalidRegistrationArgumentsException(CouldNotFindMessage("environment", missing.ToArray()));
+                if (missingByNameOnly.Any())
+                    throw new InvalidRegistrationArgumentsException(
+                        CouldNotFindByNameMessage("environment", missingByNameOnly.ToArray()));
```

Comment:
As mentioned in the description, if there is any `EnvironmentNames` value that can't be found by name, then we throw an exception, which maintains existing functionality